sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -92,6 +92,36 @@ object PhysicalOperation extends PredicateHelper {
           .map(Alias(_, a.name)(a.exprId, a.qualifier, isGenerated = a.isGenerated)).getOrElse(a)
     }
   }
+
+  /**
+   * Drop the non-partition key expression in the disjunctions, to optimize the partition pruning.
```

Comment:
Oh, OK, originally, I think the conjunction cases was handled in `collectProjectsAndFilters` already, before being passed into this function, and here, we only handle the `AND` in the disjunction. (You can see this in HiveTableScans in HiveStrategies.scala)

Anyway, you convinced me. :)